### ISSUE TITLE
Label leaked (unfair setting)

### ISSUE THREAD

**Issue by kltrock** [PhilipChicco/CADfewshot#1]
https://github.com/PhilipChicco/CADfewshot/blob/2f31f94a5c879c6f93628cfb092d871291014c96/methods/cad.py#L29

Hi,

I noticed in your CAD modules that the attentions are performed for each class independently. This requires you to know the label for the query samples in advance, which is obviously unfair.

As shown in your [code](https://github.com/PhilipChicco/CADfewshot/blob/2f31f94a5c879c6f93628cfb092d871291014c96/methods/cad.py#L39),  you have performed matrix multiplication class-wise.

**Comment by LoveMiki** [comment]
> https://github.com/PhilipChicco/CADfewshot/blob/2f31f94a5c879c6f93628cfb092d871291014c96/methods/cad.py#L29
> 
> Hi,
> 
> I noticed in your CAD modules that the attentions are performed for each class independently. This requires you to know the label for the query samples in advance, which is obviously unfair.
> 
> As shown in your [code](https://github.com/PhilipChicco/CADfewshot/blob/2f31f94a5c879c6f93628cfb092d871291014c96/methods/cad.py#L39), you have performed matrix multiplication class-wise.

I don't think so, the label is not used in MAD block. Actually, their method seems similar to CAN, which uses cross attention without knowing the labels of query samples.

**Comment by kltrock** [comment]
https://github.com/PhilipChicco/CADfewshot/blob/2f31f94a5c879c6f93628cfb092d871291014c96/methods/cad.py#L76

In the code, the query samples are sorted by class. z_query has shape (5,15,640), which means there are 5 classes, and each class has 15 query samples. 

https://github.com/PhilipChicco/CADfewshot/blob/2f31f94a5c879c6f93628cfb092d871291014c96/methods/cad.py#L39

In this line of code, Q_ has shape (5, 15, 640) and K_ has shape (5, 1, 640), where the first dimension indicates classes. The line performs batch matrix multiplication, which implies the multiplication is classwise. For example, query samples of class #1 only attend to support samples of class #1 and vice versa.

**Comment by kltrock** [comment]
In CAN, a particular query sample is cross-attended to every class in the support set, which leads to 5 versions of the same query samples. Hence, in this case, the setting is fair.

**Comment by LoveMiki** [comment]
> https://github.com/PhilipChicco/CADfewshot/blob/2f31f94a5c879c6f93628cfb092d871291014c96/methods/cad.py#L76
> 
> In the code, the query samples are sorted by class. z_query has shape (5,15,640), which means there are 5 classes, and each class has 15 query samples.
> 
> https://github.com/PhilipChicco/CADfewshot/blob/2f31f94a5c879c6f93628cfb092d871291014c96/methods/cad.py#L39
> 
> In this line of code, Q_ has shape (5, 15, 640) and K_ has shape (5, 1, 640), where the first dimension indicates classes. The line performs batch matrix multiplication, which implies the multiplication is classwise. For example, query samples of class #1 only attend to support samples of class #1 and vice versa.

I think you are right. I didn't run the code just took a glimpse on the code. Basically, the query samples should in the form of (75, 640) instead of (5, 15, 75) since we don't know the sequence of the query samples. By shaping the query samples like that, the network alreadly knows that the first 15 samples belong to the same class.

**Comment by Fu-Zhou-2019** [comment]
YES, the setting is unfair. I run the code, when I shuffle the sequence of query, the model is non-converge.
